src/openmmio.py

```diff
@@ -1117,34 +1168,68 @@ def normal_modes(self, shot=0, optimize=True):
+            Disable virtual sites (needed for SMIRNOFF)
+        include_restraint_energy : bool
+            Include energy component from CustomExternalForce
 
-        """ Optimize the geometry and align the optimized geometry to the starting geometry, and return the RMSD. """
+        Returns
+        -------
+        E : float
+            Potential energy of the system
+        rmsd : float
+            RMSD of the system (w/r.t. starting geometry) in Angstrom
+        """
 
         steps = int(max(1, -1*np.log10(crit)))
         self.update_simulation()
         self.set_positions(shot)
+        if hasattr(self, 'restrain_k'):
+            self.set_restraint_positions(shot)
         # Get the previous geometry.
         X0 = np.array([j for i, j in enumerate(self.simulation.context.getState(getPositions=True).getPositions().value_in_unit(angstrom)) if self.AtomMask[i]])
+        # printcool_dictionary(energy_components(self.simulation), title='Energy component analysis before minimization, shot %i' % shot)
         # Minimize the energy.  Optimizer works best in "steps".
         for logc in np.linspace(0, np.log10(crit), steps):
-            self.simulation.minimizeEnergy(tolerance=10**logc*kilojoule/mole)
+            self.simulation.minimizeEnergy(tolerance=10**logc*kilojoule/mole, maxIterations=10000)
+        # Remove the restraint energy from the total energy if desired.
+        groups = set(range(32))
+        if hasattr(self, 'restrain_k') and not include_restraint_energy:
+            for frc in self.simulation.system.getForces():
+                if isinstance(frc, CustomExternalForce):
+                    groups.remove(frc.getForceGroup())
+        # printcool_dictionary(energy_components(self.simulation), title='Energy component analysis after minimization, shot %i' % shot)
+        S = self.simulation.context.getState(getPositions=True, getEnergy=True, groups=groups)
         # Get the optimized geometry.
-        S = self.simulation.context.getState(getPositions=True, getEnergy=True)
         X1 = np.array([j for i, j in enumerate(S.getPositions().value_in_unit(angstrom)) if self.AtomMask[i]])
         E = S.getPotentialEnergy().value_in_unit(kilocalorie_per_mole)
         # Align to original geometry.
         M = deepcopy(self.mol[0])
         M.xyzs = [X0, X1]
-        if not self.pbc:
+        if not self.pbc and align:
             M.align(center=False)
         X1 = M.xyzs[1]
         # Set geometry in OpenMM, requires some hoops.
         mod = Modeller(self.pdb.topology, [Vec3(i[0],i[1],i[2]) for i in X1]*angstrom)
-        mod.addExtraParticles(self.forcefield)
-        # self.simulation.context.setPositions(ResetVirtualSites(mod.getPositions(), self.system))
-        self.simulation.context.setPositions(ResetVirtualSites_fast(mod.getPositions(), self.vsinfo))
-        return E, M.ref_rmsd(0)[1]
+        if disable_vsite:
+            self.simulation.context.setPositions(mod.getPositions())
```

Comment:
I think this might be improved by
```
if disable_vsite:
    self.simulation.context.setPositions(X1 * angstrom)
else:
    mod = Modeller(self.pdb.topology, X1*angstrom)
    mod.addExtraParticles(self.forcefield)
    self.simulation.context.setPositions(ResetVirtualSites_fast(mod.getPositions(), self.vsinfo))
```
In this way we can skip creating the `mod` object when `disable_vsite==True`.

Reply:
Changes have been made in the latest commit.